src/resource.ts

```diff
@@ -0,0 +1,207 @@
+        this.expiresIn &&
+        this.expiresAt &&
+        this.patient &&
+        this.refreshToken &&
+        this.scope &&
+        this.tokenType
+      )
+    ) {
+      throw new Error("Invalid authorization token.");
+    }
+  }
+}
+
+function sleep(time: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, time);
+  });
+}
+
+function isRetryable(error: any) {
+  let retryOK = false;
+
+  if (
+    error.response &&
+    retrySettings.retryableCodes.includes(error.response.status)
+  ) {
+    if (error.request && error.request.path) {
+      retryOK = endsWith(error.request.path);
+    }
+  }
+  return retryOK;
+}
+
+function endsWith(path: string) {
+  let matched = false;
+  for (let i = 0; i < retrySettings.retryableFhirPath.length; i++) {
+    if (path.endsWith(retrySettings.retryableFhirPath[i])) {
+      matched = true;
+      break;
+    }
+  }
+  return matched;
+}
+
+async function doRetry(fhirUrl: string, config: any) {
+  const interval = retrySettings.initInterval;
+  const maxAttempts = retrySettings.maxAttempts;
+
+  let resp = null;
+
+  for (let i = 0; i < maxAttempts; i += 1) {
+    const waitInMilliSec = eval(retrySettings.backOffExpr);
+    await sleep(waitInMilliSec);
+    try {
+      resp = await axios.get(fhirUrl, config);
+      break;
+    } catch (error: any) {
+      if (error.response) {
+        resp = error.response;
+      }
+    }
+  }
+  return resp;
+}
+
+async function refreshAccessToken(
+  authToken: AuthorizationToken,
+  bb: BlueButton
+) {
+  const resp = await axios.post(
+    `${String(bb.baseUrl)}/v${bb.version}/o/token/`,
+    {},
+    {
+      auth: {
+        username: bb.clientId,
+        password: bb.clientSecret,
+      },
+      params: {
+        grant_type: "refresh_token",
+        client_id: bb.clientId,
+        refresh_token: authToken.refreshToken,
+      },
+    }
+  );
+
+  if (resp.status !== 200) {
+    const msg = `Failed to refresh access token, status: ${resp.status}, error: ${resp.data}.`;
+    throw new Error(msg);
+  }
+
+  return new AuthorizationToken(resp.data);
+}
+
+export async function fetchData(
+  resourceType: FhirResourceType,
+  authToken: AuthorizationToken,
+  bb2: BlueButton,
+  queryParams: any
+) {
+  authToken.validate();
+
+  let newAuth = authToken;
+
+  if (authToken.isExpired()) {
+    newAuth = await refreshAccessToken(authToken, bb2);
+  }
+
+  let fhirUrl = `${String(bb2.baseUrl)}/v${bb2.version}`;
+
+  switch (resourceType) {
+    case FhirResourceType.Patient:
+      fhirUrl = `${fhirUrl}/fhir/Patient/`;
+      break;
+    case FhirResourceType.Coverage:
+      fhirUrl = `${fhirUrl}/fhir/Coverage/`;
+      break;
+    case FhirResourceType.ExplanationOfBenefit:
+      fhirUrl = `${fhirUrl}/fhir/ExplanationOfBenefit/`;
+      break;
+    case FhirResourceType.Profile:
+      fhirUrl = `${fhirUrl}/connect/userinfo`;
+      break;
+    default:
+      throw Error(`Unknown Fhir Resource Type --> ${resourceType}`);
+  }
```

Comment:
Related to the previous `resourceType` -> `resourcePath` a check wouldn't be needed here. If a wrong resource path is used, an Axios error would bubble up tot he SDK user.

Reply:
thinking about enum to path change, and as a SDK, one of its added value to to prevent client code to generate invalid requests to the backend as much as possible, in that regard, letting malformed (or not found resource) url going to the backend is not preferred practice, should catch the malformed requests as early as possible in SDK code  ...

the enum type for fhir resources BB2 support are fully under BB2's control, and enum provides a good mechanism to prevent mis-use when client code could not specify a resource type that is not in the enumerated types, BB2 also own the SDK, so when there is new resource added or removed or changed its URL, BB2 would make changes in SDK, that also shields client code from such changes.

just my 2c.

Reply:
I agree that we shouldn't let the user of the SDK pass a url in directly (at least most of the time). But I also think that this kind of functionality belongs at the top level, not inside this function. I feel like we should make the functions simpler. Then the wrapper class knows how to call them in a smart way. That way the underlying functions remain more flexible in the future. If you want to make this just accept a url, in my wrapping it all up story, I'll make sure to add the proper functions to the exported class.